usr/lib/cca_stdll/cca_specific.c

```diff
@@ -755,6 +755,174 @@ static CK_RV cca_get_adapter_serial_number(char *serialno)
     return CKR_OK;
 }
 
+static CK_RV cca_get_mk_state(enum cca_mk_type mk_type,
+                              enum cca_cmk_state *cur,
+                              enum cca_nmk_state *new)
+{
+    unsigned char rule_array[256] = { 0, };
+    long return_code, reason_code, rule_array_count, verb_data_length;
+    char *cmk, *nmk;
+
+    switch (mk_type) {
+    case CCA_MK_SYM:
+        memcpy(rule_array, "STATCCAE", CCA_KEYWORD_SIZE);
+        rule_array_count = 1;
+        break;
+    case CCA_MK_AES:
+        memcpy(rule_array, "STATAES ", CCA_KEYWORD_SIZE);
+        rule_array_count = 1;
+        break;
+    case CCA_MK_APKA:
+        memcpy(rule_array, "STATAPKA", CCA_KEYWORD_SIZE);
+        rule_array_count = 1;
+        break;
+    default:
+        return CKR_ARGUMENTS_BAD;
+    }
+
+    verb_data_length = 0;
+    dll_CSUACFQ(&return_code, &reason_code,
+                NULL, NULL,
+                &rule_array_count, rule_array,
+                &verb_data_length, NULL);
+
+    if (return_code != CCA_SUCCESS) {
+        TRACE_ERROR("CSUACFQ (%s) failed. return:%ld, reason:%ld\n",
+                    rule_array, return_code, reason_code);
+        return CKR_FUNCTION_FAILED;
+    }
+
+    switch (mk_type) {
+    case CCA_MK_SYM:
+        cmk = (char *)&rule_array[CCA_STATCCAE_SYM_CMK_OFFSET];
+        nmk = (char *)&rule_array[CCA_STATCCAE_SYM_NMK_OFFSET];
+        break;
+    case CCA_MK_AES:
+        cmk = (char *)&rule_array[CCA_STATAES_AES_CMK_OFFSET];
+        nmk = (char *)&rule_array[CCA_STATAES_AES_NMK_OFFSET];
+        break;
+    case CCA_MK_APKA:
+        cmk = (char *)&rule_array[CCA_STATAPKA_APKA_CMK_OFFSET];
+        nmk = (char *)&rule_array[CCA_STATAPKA_APKA_NMK_OFFSET];
+        break;
+    default:
+        return CKR_ARGUMENTS_BAD;
+    }
+
+    cmk[1] = '\0';
+    nmk[1] = '\0';
+
+    if (cur != NULL) {
+        if (sscanf(cmk, "%d", (int *)cur) != 1) {
+            TRACE_ERROR("Bad CMK status '%s'\n", cmk);
+            return CKR_FUNCTION_FAILED;
+        }
+    }
+
+    if (new != NULL) {
+        if (sscanf(nmk, "%d", (int *)new) != 1) {
+            TRACE_ERROR("Bad CMK status '%s'\n", nmk);
+            return CKR_FUNCTION_FAILED;
+        }
+    }
+
+    return CKR_OK;
+}
+
+static CK_RV cca_get_mkvps(unsigned char *cur_sym, unsigned char *new_sym,
+                           unsigned char *cur_aes, unsigned char *new_aes,
+                           unsigned char *cur_apka, unsigned char *new_apka)
+{
+    unsigned char rule_array[256] = { 0, };
```

Comment:
there's a constant for CCA_RULE_ARRAY_SIZE

Reply:
fixed